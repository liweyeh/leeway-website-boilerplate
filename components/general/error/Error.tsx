import styles from './error.module.scss';
interface ErrorProps {
	statusCode?: string;
}

const getError = (status?: string) => {
	switch (status) {
		case '404':
			return {
				code: '404',
				title: 'Error 404',
				subtext: 'There was an error with your request.',
			};
		case '500':
			return {
				code: '500',
				title: 'Internal server error',
				subtext: 'The server encountered an error and was unable to complete your request',
			};
		default:
			return {
				code: '???',
				title: 'Unknown error',
				subtext: 'An unknown error has occurred, Try refreshing the page or try again later.',
			};
	}
};

export const Error = ({ statusCode }: ErrorProps) => {
	const curErr = getError(statusCode);
	return (
		<div className={styles.error}>
			<div className={styles['error-icon']}>
				<div className={styles['error-icon-header']}>
					<div className={styles['error-icon-circle']} />
					<div className={styles['error-icon-circle']} />
					<div className={styles['error-icon-circle']} />
					<div className={styles['error-icon-triangle']}>
						<p className={styles['error-icon-triangle-text']}>!</p>
					</div>
				</div>
				<div className={styles['error-icon-text']}>
					<p>{curErr.code}</p>
				</div>
			</div>
			<p className={styles['error-title']}>{curErr.title}</p>
			<p className={styles['error-subtext']}>{curErr.subtext}</p>
		</div>
	);
};
