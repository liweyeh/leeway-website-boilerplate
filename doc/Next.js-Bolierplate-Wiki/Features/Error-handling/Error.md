[[_TOC_]]

# Overview

- Purpose: Fallback UI when there is a HTTP request error
- Type: Error Handling
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/error)
- Created by: Liwei Yeh <liwei.yeh@adasasistemas.com>
- Created time: 15/08/2022

# Props

- statusCode<`string`>: HTTP status code

# Usage

```js
const DefaultError = () => {
	const [status, setStatus] = useState('');

	const ErrorDemo = ({ status }: { status: string }) => {
		useEffect(() => {
			if (status) throw 'api error';
		}, [status]);

		return <h3> Everything is fine </h3>;
	};

	const fallback = <Error statusCode={status} />;

	return (
		<>
			<main className={styles.main}>
				<h1>Organised error handling</h1>
				<Select
					customStyles={styles.select}
					label={'Error Code:'}
					options={['404', '500']}
					handleChange={v => {
						setStatus(v);
					}}
				/>
				<ErrorBoundary fallback={fallback}>
					<ErrorDemo status={status} />
				</ErrorBoundary>
			</main>
		</>
	);
};
```

# Customisation

You can add/change the status error response by changing the following code in `Error.tsx`:

```ts
const getError = (status?: string) => {
	switch (status) {
		case '404':
			return {
				code: '404',
				title: 'Error 404',
				subtext: 'There was an error with your request.',
			};
		case '500':
			return {
				code: '500',
				title: 'Internal server error',
				subtext: 'The server encountered an error and was unable to complete your request',
			};
		default:
			return {
				code: '???',
				title: 'Unknown error',
				subtext: 'An unknown error has occurred, Try refreshing the page or try again later.',
			};
	}
};
```
