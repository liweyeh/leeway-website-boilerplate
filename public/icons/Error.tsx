import { Icon } from '@types';

export const Error = ({ customStyles, handleClick }: Icon) => {
	return (
		<svg
			className={customStyles}
			onClick={handleClick}
			xmlns="http://www.w3.org/2000/svg"
			width="18"
			height="18"
			viewBox="0 0 18 18"
		>
			<g id="Group_4217" data-name="Group 4217" transform="translate(-23384 -8449)">
				<circle
					id="Ellipse_664"
					data-name="Ellipse 664"
					cx="9"
					cy="9"
					r="9"
					transform="translate(23384 8449)"
				/>
				<g id="Group_4216" data-name="Group 4216" transform="translate(23388.689 8453.689)">
					<path
						id="Path_3305"
						data-name="Path 3305"
						d="M949.44,617.18a.615.615,0,0,1,0,.871l-7.389,7.389a.616.616,0,0,1-.871-.871l7.389-7.389A.615.615,0,0,1,949.44,617.18Z"
						transform="translate(-941 -617)"
						fill="#fff"
						fillRule="evenodd"
					/>
					<path
						id="Path_3306"
						data-name="Path 3306"
						d="M941.18,617.18a.615.615,0,0,1,.871,0l7.389,7.389a.616.616,0,0,1-.871.871l-7.389-7.389A.615.615,0,0,1,941.18,617.18Z"
						transform="translate(-941 -617)"
						fill="#fff"
						fillRule="evenodd"
					/>
				</g>
			</g>
		</svg>
	);
};
