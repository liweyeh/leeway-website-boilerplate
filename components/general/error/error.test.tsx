import { render } from '@testing-library/react';
import { Error } from './Error';
import '@testing-library/jest-dom';

const setup = (status: string) => {
	const { asFragment } = render(<Error statusCode={status} />);
	return { asFragment };
};

test('test Error 404 snapshot', () => {
	const { asFragment } = setup('404');
	expect(asFragment()).toMatchSnapshot();
});

test('test Error 500 snapshot', () => {
	const { asFragment } = setup('500');
	expect(asFragment()).toMatchSnapshot();
});
