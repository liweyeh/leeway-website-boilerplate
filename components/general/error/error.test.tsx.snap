// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Error 404 snapshot 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <div
      class="error-icon"
    >
      <div
        class="error-icon-header"
      >
        <div
          class="error-icon-circle"
        />
        <div
          class="error-icon-circle"
        />
        <div
          class="error-icon-circle"
        />
        <div
          class="error-icon-triangle"
        >
          <p
            class="error-icon-triangle-text"
          >
            !
          </p>
        </div>
      </div>
      <div
        class="error-icon-text"
      >
        <p>
          404
        </p>
      </div>
    </div>
    <p
      class="error-title"
    >
      Error 404
    </p>
    <p
      class="error-subtext"
    >
      There was an error with your request.
    </p>
  </div>
</DocumentFragment>
`;

exports[`test Error 500 snapshot 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <div
      class="error-icon"
    >
      <div
        class="error-icon-header"
      >
        <div
          class="error-icon-circle"
        />
        <div
          class="error-icon-circle"
        />
        <div
          class="error-icon-circle"
        />
        <div
          class="error-icon-triangle"
        >
          <p
            class="error-icon-triangle-text"
          >
            !
          </p>
        </div>
      </div>
      <div
        class="error-icon-text"
      >
        <p>
          500
        </p>
      </div>
    </div>
    <p
      class="error-title"
    >
      Internal server error
    </p>
    <p
      class="error-subtext"
    >
      The server encountered an error and was unable to complete your request
    </p>
  </div>
</DocumentFragment>
`;
